utilities/write_batch_with_index/write_batch_with_index.cc

```diff
@@ -329,14 +412,105 @@ class BaseDeltaIterator : public Iterator {
 #endif  // __clang_analyzer__
   }
 
-  bool forward_;
+  /**
+   * Returns the upper bound for the base iterator,
+   * or nullptr if there is no upper bound.
+   *
+   * The base iterator may have its own upper bound,
+   * if not not we use the upper bound from this
+   * iterator's ReadOptions (if present).
+   */
+  inline const Slice* base_iterator_upper_bound() const {
+    const Slice* upper_bound = base_iterator_->upper_bound();
+    if (upper_bound == nullptr && read_options_ != nullptr) {
+      return read_options_->iterate_upper_bound;
+    }
+    return upper_bound;
+  }
+
+  /**
+   * Returns the lower bound for the base iterator,
+   * or nullptr if there is no lower bound.
+   *
+   * The base iterator may have its own lower bound,
+   * if not not we use the lower bound from this
+   * iterator's ReadOptions (if present).
+   */
+  inline const Slice* base_iterator_lower_bound() const {
+    const Slice* lower_bound = base_iterator_->lower_bound();
+    if (lower_bound == nullptr && read_options_ != nullptr) {
+      return read_options_->iterate_lower_bound;
+    }
+    return lower_bound;
+  }
+
+  bool BaseIsWithinBounds() const {
+    const Slice* lower_bound = base_iterator_lower_bound();
```

Comment:
Seems for efficiency we should check IsMovingBackward() before looking up lower bound. (Same for forward and upper bound.)

Reply:
Thanks @pdillinger that's a good suggestion. I have now pushed an additional commit for that.